sdk/src/azure/core/az_mqtt5_rpc_client_codec.c

```diff
@@ -1,83 +1,420 @@
+  int32_t command_id_index = az_span_find(topic_format, az_mqtt5_rpc_command_id_key);
+
+  int32_t required_length = az_span_size(topic_format) + (int32_t)sizeof((uint8_t)'\0');
+  required_length += (invoker_client_index == -1)
+      ? 0
+      : az_span_size(client->_internal.client_id) - az_span_size(az_mqtt5_rpc_client_id_key);
+  required_length += (model_id_index == -1)
+      ? 0
+      : az_span_size(client->_internal.model_id) - az_span_size(az_mqtt5_rpc_service_id_key);
+  required_length += (executor_id_index == -1)
+      ? 0
+      : (int32_t)sizeof((uint8_t)'+') - az_span_size(az_mqtt5_rpc_executor_id_key);
+  required_length += (command_id_index == -1)
+      ? 0
+      : (int32_t)sizeof((uint8_t)'+') - az_span_size(az_mqtt5_rpc_command_id_key);
+
+  _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, required_length);
+
+  az_span remainder = mqtt_topic_span;
+
+  for (int32_t i = 0; i < az_span_size(topic_format); i++)
+  {
+    uint8_t c = az_span_ptr(topic_format)[i];
+
+    if (c == '{')
+    {
+      if (i == invoker_client_index)
+      {
+        remainder = az_span_copy(remainder, client->_internal.client_id);
+        i += az_span_size(az_mqtt5_rpc_client_id_key) - 1;
+      }
+      else if (i == model_id_index)
+      {
+        remainder = az_span_copy(remainder, client->_internal.model_id);
+        i += az_span_size(az_mqtt5_rpc_service_id_key) - 1;
+      }
+      else if (i == executor_id_index)
+      {
+        remainder = az_span_copy_u8(remainder, '+');
+        i += az_span_size(az_mqtt5_rpc_executor_id_key) - 1;
+      }
+      else if (i == command_id_index)
+      {
+        remainder = az_span_copy_u8(remainder, '+');
+        i += az_span_size(az_mqtt5_rpc_command_id_key) - 1;
+      }
+    }
+    else
+    {
+      remainder = az_span_copy_u8(remainder, c);
+    }
+  }
+
+  az_span_copy_u8(remainder, '\0');
+
+  if (out_mqtt_topic_length)
+  {
+    *out_mqtt_topic_length = (size_t)required_length;
+  }
+
+  return AZ_OK;
 }
 
-AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_get_request_topic(
+AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_parse_received_topic(
     az_mqtt5_rpc_client_codec* client,
-    az_span server_client_id,
-    az_span command_name,
-    az_span out_request_topic)
+    az_span received_topic,
+    az_mqtt5_rpc_client_codec_request_response* out_response)
 {
-  return az_rpc_get_topic_from_format(
-      client->_internal.options.request_topic_format,
-      client->_internal.model_id,
-      server_client_id,
-      client->_internal.client_id,
-      _az_span_is_valid(client->_internal.command_name, 1, 0) ? client->_internal.command_name
-                                                              : command_name,
-      out_request_topic,
-      NULL);
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  _az_PRECONDITION_NOT_NULL(out_response);
+
+  az_span topic_format = client->_internal.options.subscription_topic_format;
```

Comment:
This is the same code as above - it should be refactored so that the same function is reused.

Reply:
Will implement in a follow-up PR